File: data-plane/python/integrations/slimrpc/slimrpc/context.py
# Copyright AGNTCY Contributors (https://github.com/agntcy)
# SPDX-License-Identifier: Apache-2.0

from dataclasses import dataclass

import slim_bindings


@dataclass
class SessionContext:
    """
    Context for RPC calls based on session information.
    """

    session_id: int
    source_name: str
    destination_name: str
    metadata: dict[str, str] | None = None

    @classmethod
    def from_session(cls, session: slim_bindings.Session) -> "SessionContext":
        """
        Create a SessionContext from session.
        """
        return cls(
            session_id=session.id,
            source_name=str(session.src),
            destination_name=str(session.dst),
            metadata=session.metadata,
        )


@dataclass
class MessageContext:
    """
    Context for RPC calls based on message information.
    """

    source_name: str
    destination_name: str
    payload_type: str
    metadata: dict[str, str] | None = None

    @classmethod
    def from_message_context(
        cls, msg_ctx: slim_bindings.MessageContext
    ) -> "MessageContext":
        """
        Create a MessageContext from message context.
        """
        return cls(
            source_name=str(msg_ctx.source_name),
            destination_name=str(msg_ctx.destination_name),
            payload_type=msg_ctx.payload_type,
            metadata=msg_ctx.metadata,
        )


# Keep Context as an alias for backward compatibility
Context = SessionContext

File: data-plane/core/slimrpc/src/context.rs
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

use slim_datapath::messages::Name;
use slim_session::session_controller::SessionController;
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct SessionContext {
    pub session_id: u64,
    pub source_name: Name,
    pub destination_name: Name,
    pub metadata: Option<HashMap<String, String>>,
}

impl SessionContext {
    pub fn from_session(session: &Arc<SessionController>) -> Self {
        Self {
            session_id: session.id() as u64,
            source_name: session.source().clone(),
            destination_name: session.dst().clone(),
            metadata: Some(session.metadata()),
        }
    }
}

#[derive(Debug, Clone)]
pub struct MessageContext {
    pub source_name: Name,
    pub destination_name: Name,
    pub payload_type: String,
    pub metadata: Option<HashMap<String, String>>,
}

impl MessageContext {
    pub fn from_message(msg: &slim_datapath::api::ProtoMessage) -> Self {
        Self {
            source_name: msg.get_slim_header().get_source(),
            destination_name: msg.get_slim_header().get_dst(),
            payload_type: "application".to_string(), // TODO: Get from message
            metadata: Some(msg.get_metadata_map()),
        }
    }
}

// Keep Context as alias for backward compatibility
pub type Context = SessionContext;
